File: 139-word-break/139-word-break.java
class Node {
    char ch;
    Node[] link;
    boolean isEnd = false;
    
    Node(char ch) {
        this.ch = ch;
        this.link = new Node[26];
    }
}

class TrieOperations {
    Node root;
    
    TrieOperations() {
        root = new Node('\0');
    }
    
    public void insert(String word) {
        Node curr = root;
        for(int i=0;i<word.length();i++) {
            char c = word.charAt(i);
            if(curr.link[c - 'a'] == null) curr.link[c - 'a'] = new Node(c);
            curr = curr.link[c - 'a'];
        }
        curr.isEnd = true;
    }
}

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        TrieOperations trie = new TrieOperations();
        
        for(String word : wordDict) trie.insert(word);
        
        boolean[] dp = new boolean[s.length() + 1];
        char[] str = s.toCharArray();
        dp[0] = true;
        
        for(int i=0;i<str.length;i++) {
            if(!dp[i]) continue;
            
            int j = i;
            Node curr = trie.root;
            while(j < str.length && curr.link[str[j] - 'a'] != null) {
                curr = curr.link[str[j++] - 'a'];
                if(curr.isEnd) {
                    dp[j] = true;
                }
            }
        }
        return dp[str.length];
    }
}

File: 139-word-break/139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.length();
        bool dp[n+1][n+1];
        
        unordered_set<string> dict;
        
        for(string word : wordDict) dict.insert(word);
        
        for(int i=1;i<=n;++i) { //window size
            int start = 1;
            int end = i;
            
            while(end <= n) { //sliding window
                string sub = s.substr(start - 1, i);
                if(dict.find(sub) != dict.end()) dp[start][end] = true;
                else { //partition into smaller and check
                    bool flag = false;
                    for(int p=start;p<end;++p) {
                        if(dp[start][p] && dp[p + 1][end]) {
                            flag = true;
                            break;
                        }
                    }
                    dp[start][end] = flag;
                }
                start++;
                end++;
            }
        }
        return dp[1][n];
    }
};